drivers: rtc: add silabs siwx91x rtc driver

Martinhoff-maker opened the pull request:
The goal of this PR is to add a basic RTC driver for the Silicon Labs SiWx_91x SoC (with tests).

One rtc_api test is currently failing at runtime due to this issue: https://github.com/zephyrproject-rtos/zephyr/pull/89610

Clock control is somewhat tricky because the driver can function without explicit clock control, as sl_calendar_init() from the WiseConnect HAL handles the initialization. However, we could potentially use sl_calendar_rtc_start() to avoid relying on the HAL for clock initialization. This approach would require implementing a wait loop in the clock control code because of this:

In this code, sl_calendar_init() is equivalent to RSI_RTC_Init() and sl_calendar_rtc_start() is equivalent to RSI_RTC_Start()

```
void RSI_RTC_Init(RTC_Type *Cal)
{
  /* Power-Up RTC Domain */
  RSI_PS_NpssPeriPowerUp(SLPSS_PWRGATE_ULP_MCURTC | SLPSS_PWRGATE_ULP_TIMEPERIOD);

  /*Wait for RTC-ON pulse synchronization*/
  do {
    RTC->MCU_CAL_ALARM_PROG_1 = 6U;
  } while (RTC->MCU_CAL_ALARM_PROG_1 != 6U);
  // by using this API we programmed the RTC timer clock in SOC
  // MSB 8-bits for the Integer part &
  // LSB 17-bits for the Fractional part
  // Eg:- 32KHz = 31.25µs ==> 31.25*2^17 = 4096000 = 0x3E8000
  /* Time Period Programming */
  RSI_TIMEPERIOD_TimerClkSel(TIME_PERIOD, 0x003E7FFF);
  /*Enable power gates*/
  RSI_RTC_Start(Cal);
  return;
}
```

jhedberg commented:
@Martinhoff-maker what's the reason the HAL PR dependency is still a Draft, but this PR is not?

Martinhoff-maker commented:
@jhedberg I forget to open it. I've open the HAL PR now.

bjarki-andreasen reviewed (dismissed):
Looks fine :) Remember to update to sha in west.yml

jerome-pouiller commented on drivers/rtc/Kconfig.siwx91x:
```diff
@@ -0,0 +1,9 @@
+	default y
+	depends on DT_HAS_SILABS_SIWX91X_RTC_ENABLED
+	help
+	  Build RTC driver for Silabs SIWX91X soc
```
The Reference Manual talk about "Calendar hardware block" while Zephyr talk about "RTC". In addition, there is also another clock called "RTC" on SiWx91x. Maybe you can mention that in the help section (or in the dt-bindings).

jerome-pouiller commented on drivers/rtc/rtc_silabs_siwx91x.c:
```diff
@@ -0,0 +1,153 @@
+
+static int silabs_siwx91x_rtc_set_time(const struct device *dev, const struct rtc_time *timeptr)
+{
+	sl_calendar_datetime_config_t siwx91x_time = {0};
```
`0` is useless. You can use `= { };`

Martinhoff-maker replied:
done

jerome-pouiller commented on drivers/rtc/rtc_silabs_siwx91x.c:
```diff
@@ -0,0 +1,153 @@
+
+static int silabs_siwx91x_rtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	sl_calendar_datetime_config_t siwx91x_time = {0};
```
Ditto

jerome-pouiller commented on drivers/rtc/rtc_silabs_siwx91x.c:
```diff
@@ -0,0 +1,153 @@
+	return ret;
+}
+
+static int silabs_siwx91x_rtc_get_time(const struct device *dev, struct rtc_time *timeptr)
```
Some symbols are prefixed by `silabs_siwx91x_rtc_`, some others by `siwx91x_rtc_` and there is also `rtc_time_to_siwx91x_time_set()` and `siwx91x_time_to_rtc_time_set()`. Can you uniformize the names?

Martinhoff-maker replied:
uniformize with siwx91x_rtc_ for API function + data/config structure. rtc_time_to_siwx91x_time_set() and siwx91x_time_to_rtc_time_set() are conversion functions so it's ok IMO

jerome-pouiller reviewed (commented):
Only nitpick and documentation issues.

Martinhoff-maker committed:
drivers: rtc: add silabs siwx91x rtc driver

Add the support of silabs siwx91x basic rtc driver.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
tests: drivers: rtc: add siwg917_rb4338a board overlay and add rtc tag

Add rtc tag to the siwg917_rb4338a board to activate testing on the
board.
Also add an overlay for the board in order to run the test correctly.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker committed:
dts: arm: silabs: add rtc node for siwg917

Add rtc node that was present on siwx91x baords.

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented:
v3: rebased on top of main branch

jerome-pouiller committed:
tests: drivers: pwm: Fix build

Commit a1f48216bc7 ("soc: silabs: siwx91x: Rename pins") broke the test
for pwm_api.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

Martinhoff-maker commented:
v4: add patch for pwm in order to resolve the general CI build failure in the same time.

Martinhoff-maker commented:
@bjarki-andreasen Sorry for the multiple calls, have some issue with the CI but it's fixed now.
